### PR TITLE
Implement update methods CSV & NdJson

### LINKED ISSUE #215: NDJSON/CSV methods to add and update documents
⚠️ This issue is generated, it means the nameing might be done differently in this package (ex: `add_documents_json` instead of `addDocumentsJson`). Keep the already existing way of naming in this package to stay idiomatic with the language and this repository.

📣 We strongly recommend doing multiple PRs to solve all the points of this issue

[MeiliSearch v0.23.0](https://github.com/meilisearch/MeiliSearch/releases/tag/v0.23.1) introduces two changes:
- new valid formats to push data files, additionally to the JSON format: CSV and NDJSON formats.
- it enforces the `Content-type` header for every route requiring a payload (`POST` and `PUT` routes)

---

Here are the expected changes to completely close the issue:

- [ ] Currently, the SDKs always send `Content-Type: application/json` to every request. Only the `POST` and `PUT` requests should send the `Content-Type: application/json` and not the `DELETE` and `GET` ones.

- [ ] Add the following methods and 🔥 **the associated tests** 🔥 to ADD the documents. Depending on the format type (`csv` or `ndjson`) the SDK should send `Content-Type: application/x-dnjson` or `Content-Type: text/csv`)
  - [ ] `addDocumentsJson(string docs, string primaryKey)`
  - [ ] `addDocumentsCsv(string docs, string primaryKey)`
  - [ ] `addDocumentsCsvInBatches(string docs, int batchSize, string primaryKey)`
  - [ ] `addDocumentsNdjson(string docs, string primaryKey)`
  - [ ] `addDocumentsNdjsonInBatches(string docs, int batchSize, string primaryKey)`

- [ ] Add the following methods and 🔥 **the associated tests** 🔥 to UPDATE the documents. Depending on the format type (`csv` or `ndjson`) the SDK should send `Content-Type: application/x-dnjson` or `Content-Type: text/csv`)
  - [ ] `updateDocumentsJson(string docs, string primaryKey)`
  - [ ] `updateDocumentsCsv(string docs, string primaryKey)`
  - [ ] `updateDocumentsCsvInBatches(string docs, int batchSize, string primaryKey)`
  - [ ] `updateDocumentsNdjson(string docs, string primaryKey)`
  - [ ] `updateDocumentsNdjsonInBatches(string docs, int batchSize, string primaryKey)`

`docs` are the documents sent as `String`
`primaryKey` is [the primary key of the index](https://docs.meilisearch.com/learn/core_concepts/indexes.html#primary-key)
`batchSize` is the size of the batch. Example: you can send 2000 documents in raw String in `docs` and ask for a `batchSize` of 1000, so your documents will be sent to MeiliSearch in two batches.

Example of PRs:
- in PHP SDK: https://github.com/meilisearch/meilisearch-php/pull/235
- in Python SDK: https://github.com/meilisearch/meilisearch-python/pull/329

----------

Related to: https://github.com/meilisearch/integration-guides/issues/146

If this issue is partially/completely implemented, feel free to let us know.

### PR DESCRIPTION
**Pull request by Azanul**
# Pull Request

## Related issue
Fixes #215 

## What does this PR do?
- [X] Add `UpdateDocumentsInBatches` to `IndexInterface`
- [X] Implement `UpdateDocumentsCsv`
- [X] Implement `UpdateDocumentsCsvInBatches`
- [X] Implement `UpdateDocumentsNdjson`
- [X] Implement `UpdateDocumentsNdjsonInBatches`
- [X] Implement respective tests
- [X] Implement `saveDocuments` function to reduce code duplication

## PR checklist
Please check if your PR fulfills the following requirements:
- [X] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [X] Have you read the contributing guidelines?
- [X] Have you made sure that the title is accurate and descriptive of the changes?

### CODE REVIEW

File: index_documents.go [hunk 1]
```diff
@@ -366,6 +370,169 @@ func (i Index) UpdateDocumentsInBatches(documentsPtr interface{}, batchSize int,
...
+	sendCsvRecords := func(records [][]string) (*TaskInfo, error) {
+		b := new(bytes.Buffer)
+		w := csv.NewWriter(b)
+		w.UseCRLF = true // Keep output RFC 4180 compliant
+		err := w.WriteAll(records)
+		if err != nil {
+			return nil, fmt.Errorf("could not write CSV records: %w", err)
+		}
+
+		resp, err := i.UpdateDocumentsCsv(b.Bytes(), primaryKey...)
```
Comment (alallema) [inline index_documents.go:406]: Except for this line, this method is the exact same as `AddDocumentsNdjsonFromReaderInBatches` maybe we can find a way to re-used the code that already existing than duplicated it? WDYT?

File: index_documents.go [hunk 2]
```diff
@@ -366,6 +370,169 @@ func (i Index) UpdateDocumentsInBatches(documentsPtr interface{}, batchSize int,
...
+			if err != nil {
+				return nil, fmt.Errorf("Could not write NDJSON line: %w", err)
+			}
+			err = b.WriteByte('\n')
+			if err != nil {
+				return nil, fmt.Errorf("Could not write NDJSON line: %w", err)
+			}
+		}
+
+		resp, err := i.UpdateDocumentsNdjson(b.Bytes(), primaryKey...)
```
Comment (alallema) [inline index_documents.go:488]: Same for this method maybe a trick exist to avoid the duplication of this piece of code

**Review by alallema — COMMENTED** [review]
Thanks @Azanul  for this PR!
I have added a few comments. Also every time you create a method or modify something, you need to create the associated tests to check it.
Don't forget to put your PR as `Ready for review` when it is ready and you can request me as a reviewer when you need a review 😊 
Screenshot 2023-02-08 at 15 03 55

File: index_documents.go (hunk 1 above)
Comment (Azanul) [reply index_documents.go:406]: Sure, I think a higher order function will resolve this.

Comment (Azanul) [reply index_documents.go:406]: > Except for this line, this method is the exact same as `AddDocumentsNdjsonFromReaderInBatches` maybe we can find a way to re-used the code that already existing than duplicated it? WDYT?

I hope you mean `AddDocumentsCsvFromReaderInBatches`

File: index_documents.go (hunk 2 above)
Comment (Azanul) [reply index_documents.go:473]: check it out, if you approve this approach, I'll use the same for other methods too where this pattern is occurring.

Comment (alallema) [reply index_documents.go:493]: It's better, but the problem is that almost all the method is duplicated from line 408 to 452 is possible to group all the duplicated code?

Comment (Azanul) [reply index_documents.go:493]: How about using this same approach to implement a template design pattern?

Comment (Azanul) [reply index_documents.go:452]: Check it out. Also, should I be doing the same for from reader functions?

File: index_documents.go [hunk 3]
```diff
@@ -70,26 +39,30 @@ func (i Index) addDocuments(documentsPtr interface{}, contentType string, primar
...
 		withResponse:        resp,
 		acceptedStatusCodes: []int{http.StatusAccepted},
-		functionName:        "AddDocuments",
+		functionName:        function + "Documents",
 	}
 	if err = i.client.executeRequest(req); err != nil {
 		return nil, err
 	}
 	return resp, nil
 }
```
Comment (alallema) [inline index_documents.go:63]: Sorry, I realized that I may not have been clear. I was not talking about refactoring the code of all the functions; for me this method (`saveDocuments`) only makes the understanding of the code more complex
I was talking about lines 367 to 398 (previously 408 to 452) which are always duplicated and could be put in the same function.

Comment (Azanul) [reply index_documents.go:63]: Check it out if that's what you meant

File: index_documents.go [hunk 4]
```diff
@@ -70,80 +39,30 @@ func (i Index) addDocuments(documentsPtr interface{}, contentType string, primar
...
```
Comment (alallema) [inline index_documents.go:63]: This method isn't needed you can call `addDocuments()` or `updateDocuments()` instead

File: index_documents.go (hunk 3 above)
Comment (alallema) [reply index_documents.go:63]: Sorry for the delay.
That's right, except you don't have to rewrite the `addDocuments()` method, we don't need the `saveDocument()`.

**Review by alallema — APPROVED** [review]
Thank you a lot @Azanul for this PR and all the modifications ✨ 
This will be a great enhancement 🔥 
LGTM! 🎉

**Pull request merged**